### PR TITLE
Fix the error of gradient of np.pad

### LINKED ISSUE #19043: [Numpy][Bug] The gradient of np.pad is wrong!
The gradient of `np.pad` is wrong. See the following reproducible example:

MXNet Implementation:

```python
import mxnet as mx
mx.npx.set_np()

ctx = mx.gpu()
a = mx.np.ones((3, 3, 3), ctx=ctx)
mult = np.random.normal(0, 1, (3, 3, 3))
a.attach_grad()
with mx.autograd.record():
    b = mx.np.pad(a[:, 1:], ((0, 0), (0, 1), (0, 0))) * mx.np.array(mult, ctx=ctx)
    b = b.sum()
b.backward()
print(a.grad)
```
Output:
```
[[[0. 0. 0.]
  [1. 1. 1.]
  [1. 1. 1.]]

 [[0. 0. 0.]
  [1. 1. 1.]
  [1. 1. 1.]]

 [[0. 0. 0.]
  [1. 1. 1.]
  [1. 1. 1.]]] @gpu(0)
```

Jax Implementation:

```python
from jax import grad
import jax.numpy as jnp
import numpy as np
mult = np.random.normal(0, 1, (3, 3, 3))

a = jnp.ones((3, 3, 3))

def f(x):
    b = jnp.pad(x[:, 1:], ((0, 0), (0, 1), (0, 0))) * jnp.array(mult)
    return b.sum()
print(grad(f)(a))

```

Output:
```
[[[ 0.          0.          0.        ]
  [ 0.3545383  -0.84326786 -0.31482664]
  [ 1.0994871  -1.230104    2.8007567 ]]

 [[ 0.          0.          0.        ]
  [ 1.0447861  -0.16119051 -0.39860427]
  [-0.7756538   0.5314936   1.4601654 ]]

 [[ 0.          0.          0.        ]
  [ 0.37878916 -2.0777514   0.96676654]
  [ 0.45230922  0.3094176  -0.43687683]]]
```

Basically, the following line is not correct:

https://github.com/apache/incubator-mxnet/blob/b0c39f7ea983639093c63d7d2486bbef083a55d6/src/operator/numpy/np_pad_op-inl.h#L544-L551

We should change that to 

```
 KERNEL_ASSIGN(out[i], req, a[i]);
```

In addition, I do not know why we need the `using namespace mxnet_op; `. @CassiniXu

### PR DESCRIPTION
**Pull request by CassiniXu**
Fixes #19043

@sxjscience

### CODE REVIEW

File: src/operator/numpy/np_pad_op-inl.h [hunk 1]
```diff
@@ -792,15 +825,50 @@ void NumpyPadOpBackward(const nnvm::NodeAttrs& attrs,
...
-  const TBlob& out_data = outputs[0];
-  NumpyPadOpBackImpl<xpu>(in_data, out_data,
-                          out_data.Size(), req, s);
+  MXNET_NDIM_SWITCH(inputs[0].ndim(), NDim, {
+    using namespace mxnet_op;
+    using namespace mshadow;
+    CHECK_EQ(inputs.size(), 1U);
+    CHECK_EQ(outputs.size(), 1U);
+    CHECK_EQ(req.size(), 1U);
+    CHECK_EQ(req[0], kWriteTo);
```
Comment (sxjscience) [inline src/operator/numpy/np_pad_op-inl.h:834]: You will need to also implement kAddTo.

**Comment by CassiniXu** [comment]
@mxnet-bot run ci [unix-cpu, windows-cpu]

**Comment by CassiniXu** [comment]
@mxnet-bot run ci [centos-cpu]

File: src/operator/numpy/np_pad_op-inl.h [hunk 2]
```diff
@@ -720,20 +729,44 @@ void NumpyPadOpImpl(const TBlob& in_data,
...
-    });
-  });
+    using namespace mxnet_op;
+    using namespace mshadow;
+    int mode = param.mode;
+    int ndim = in_data.ndim();
+    MXNET_NDIM_SWITCH(ndim, NDim, {
+      mshadow::Shape<NDim*2> width;
+      int dimcounter = 0;
+      index_t* odptr = reinterpret_cast<index_t*>(oshape);
```
Comment (sxjscience) [inline src/operator/numpy/np_pad_op-inl.h:745]: Need to investigate if it's possible to remove the reinterpret_cast.

File: src/operator/numpy/np_pad_op-inl.h [hunk 3]
```diff
@@ -720,20 +729,44 @@ void NumpyPadOpImpl(const TBlob& in_data,
...
-  using namespace mshadow;
-  MSHADOW_TYPE_SWITCH_WITH_BOOL(out_data.type_flag_, DType, {
-    MXNET_ASSIGN_REQ_SWITCH(req[0], req_type, {
-      Kernel<pad_grad<xpu, req_type>, xpu>::Launch(
-        s, dsize, out_data.dptr<DType>(), in_data.dptr<DType>());
-    });
-  });
+    using namespace mxnet_op;
+    using namespace mshadow;
+    int mode = param.mode;
```
Comment (sxjscience) [inline src/operator/numpy/np_pad_op-inl.h:740]: Need to raise an error when mode is not constant.

File: src/operator/numpy/np_pad_op-inl.h (hunk 2 above)
Comment (CassiniXu) [reply src/operator/numpy/np_pad_op-inl.h:748]: Thanks. I am trying to avoid using 'reinterpret_cast'.

Comment (CassiniXu) [reply src/operator/numpy/np_pad_op-inl.h:748]: > Need to investigate if it's possible to remove the reinterpret_cast.

Thanks. I am trying to avoid using 'reinterpret_cast'.

**Comment by sxjscience** [comment]
@ciyongch @szha Would you know if it's possible to backport the bug-fix to 1.7.x and 1.x?

**Comment by ciyongch** [comment]
Hi @sxjscience , thanks for reaching us for the issue, I'm afraid it's not able to be part of official 1.7.0 release as it's almost completed. But I think this fix should be fine to be backported into both 1.7.x and 1.x branches, and it's likely to be included in the next release (version: 1.8.0) which lead by @samskalicky .
